src/cpp/session/modules/data/DataViewer.cpp

```diff
@@ -246,6 +246,7 @@ struct CachedFrame
       }
 
       // cache number of columns
+      nrow = safeDim(sexp, DIM_ROWS);
```

Comment:
I wonder if we should only do this for `data.frame` or `matrix` objects -- for database tables, I think this could be expensive.